In [145]:
''' statements to import libraries'''
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import seaborn as sns
from IPython.display import display, HTML


In [110]:
''' import my csv files'''
white_df = pd.read_csv('wine+quality/winequality-white.csv', sep= ';')
white_copy = white_df.copy()
red_df = pd.read_csv('wine+quality/winequality-red.csv', sep = ';')
red_copy = red_df.copy()

print the first 10 and the last 10 records (use head() and tail() methods)

White: head and Tail

In [111]:
display(white_copy.head(10))
display(white_copy.tail(10))

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   
5            8.1              0.28         0.40             6.9      0.050   
6            6.2              0.32         0.16             7.0      0.045   
7            7.0              0.27         0.36            20.7      0.045   
8            6.3              0.30         0.34             1.6      0.049   
9            8.1              0.22         0.43             1.5      0.044   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   
5                 30.0                  97.0   0.9951  3.26       0.44   
6                 30.0                 136.0   0.9949  3.18       0.47   
7                 45.0                 170.0   1.0010  3.00       0.45   
8                 14.0                 132.0   0.9940  3.30       0.49   
9                 28.0                 129.0   0.9938  3.22       0.45   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6  
5     10.1        6  
6      9.6        6  
7      8.8        6  
8      9.5        6  
9     11.0        6

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4888            6.8             0.220         0.36            1.20      0.052   
4889            4.9             0.235         0.27           11.75      0.030   
4890            6.1             0.340         0.29            2.20      0.036   
4891            5.7             0.210         0.32            0.90      0.038   
4892            6.5             0.230         0.38            1.30      0.032   
4893            6.2             0.210         0.29            1.60      0.039   
4894            6.6             0.320         0.36            8.00      0.047   
4895            6.5             0.240         0.19            1.20      0.041   
4896            5.5             0.290         0.30            1.10      0.022   
4897            6.0             0.210         0.38            0.80      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4888                 38.0                 127.0  0.99330  3.04       0.54   
4889                 34.0                 118.0  0.99540  3.07       0.50   
4890                 25.0                 100.0  0.98938  3.06       0.44   
4891                 38.0                 121.0  0.99074  3.24       0.46   
4892                 29.0                 112.0  0.99298  3.29       0.54   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
4888      9.2        5  
4889      9.4        6  
4890     11.8        6  
4891     10.6        6  
4892      9.7        5  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6

red: Head and Tail

In [112]:
display(red_copy.head(10))
display(red_copy.tail(10))

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            7.5              0.50         0.36             6.1      0.071   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 17.0                 102.0   0.9978  3.35       0.80   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5  
5      9.4        5  
6      9.4        5  
7     10.0        7  
8      9.5        7  
9     10.5        5

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1589            6.6             0.725         0.20             7.8      0.073   
1590            6.3             0.550         0.15             1.8      0.077   
1591            5.4             0.740         0.09             1.7      0.089   
1592            6.3             0.510         0.13             2.3      0.076   
1593            6.8             0.620         0.08             1.9      0.068   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1589                 29.0                  79.0  0.99770  3.29       0.54   
1590                 26.0                  35.0  0.99314  3.32       0.82   
1591                 16.0                  26.0  0.99402  3.67       0.56   
1592                 29.0                  40.0  0.99574  3.42       0.75   
1593                 28.0                  38.0  0.99651  3.42       0.82   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
1589      9.2        5  
1590     11.6        6  
1591     11.6        6  
1592     11.0        6  
1593      9.5        6  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6

check the info and the shape (use info() and shape methods)
White:

In [113]:
display(white_copy.info())
display(white_copy.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


None

(4898, 12)

check the info and the shape (use info() and shape methods)
Red:

In [114]:
display(red_copy.info())
display(red_copy.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


None

(1599, 12)

<font color ='green'><span style="font-size:larger;">Summary: Each CSV file has 12 columns (indexing begins at 0). The White is larger. It has 4898 wine entries. The red has 1599.
The data is entirely numeric. Of the 12 columns 11 are floats. The final column 'quality' is an integer.</span></font>

observe if there are missing values (use isna(), isnull()[https://pandas.pydata.org/docs/reference/api/pandas.isnull.html])


White

In [115]:
display(white_copy.isna().sum())


fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

Red:

In [116]:
display(red_copy.isnull().sum())

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

<font color= 'green'><span style="font-size:larger;">Summary: Good. so we have 0 null or na entries in the data set</span></font>

check for Descriptive Statistics for both red and white wine datasets (use describe() method)

<span style="font-size: larger; color: white;">White</span>

In [117]:
display(white_copy.describe())

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

<span style="font-size: larger; color: red;">Red</span>

In [118]:
display(red_copy.describe())

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

Let's now group our qualities into 3 categories: low, medium and high quality wines.

<span style="font-size: larger; color: white;">White</span>

In [119]:
white_copy['quality_label'] = 'high'
white_copy.loc[white_copy.quality < 5, 'quality_label'] = 'low'
white_copy.loc[(white_copy.quality >= 5) & (white_copy.quality <=7), 'quality_label'] = 'medium'
#df.loc[df['existing_column'] > 15, 'new_column'] = 'high_value'

display(white_copy.head())

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality quality_label  
0      8.8        6        medium  
1      9.5        6        medium  
2     10.1        6        medium  
3      9.9        6        medium  
4      9.9        6        medium

<span style="font-size: larger; color: red;">Red</span>

In [120]:
red_copy['quality_label'] = red_copy.quality.apply(lambda value: 'low' if value <= 5 else 'medium' if value <= 7 else 'high')
display(red_copy.head())

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality quality_label  
0      9.4        5           low  
1      9.8        5           low  
2      9.8        5           low  
3      9.8        6        medium  
4      9.4        5           low

Create a new Dataframe combining both White and Red. White_red_df

In [121]:
white_red_df = pd.concat([white_copy, red_copy])
display(white_red_df.shape)
print(white_red_df.head())

(6497, 13)

   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality quality_label  
0      8.8        6        medium  
1      9.5    

Using the red_wine and white_wine datasets,create a table like the one below to compare Descriptive Statistics between red and white wines.


In [122]:
display(pd.concat([white_copy.describe(), red_copy.describe()],axis=1, keys = ['White wine', 'Red wine']))

White wine                                                            \
      fixed acidity volatile acidity  citric acid residual sugar    chlorides   
count   4898.000000      4898.000000  4898.000000    4898.000000  4898.000000   
mean       6.854788         0.278241     0.334192       6.391415     0.045772   
std        0.843868         0.100795     0.121020       5.072058     0.021848   
min        3.800000         0.080000     0.000000       0.600000     0.009000   
25%        6.300000         0.210000     0.270000       1.700000     0.036000   
50%        6.800000         0.260000     0.320000       5.200000     0.043000   
75%        7.300000         0.320000     0.390000       9.900000     0.050000   
max       14.200000         1.100000     1.660000      65.800000     0.346000   

                                                                          \
      free sulfur dioxide total sulfur dioxide      density           pH   
count         4898.000000          4898.000000  4898.000000  4898.000000   
mean            35.308085           138.360657     0.994027     3.188267   
std             17.007137            42.498065     0.002991     0.151001   
min              2.000000             9.000000     0.987110     2.720000   
25%             23.000000           108.000000     0.991723     3.090000   
50%             34.000000           134.000000     0.993740     3.180000   
75%             46.000000           167.000000     0.996100     3.280000   
max            289.000000           440.000000     1.038980     3.820000   

                    ...     Red wine                              \
         sulphates  ...  citric acid residual sugar    chlorides   
count  4898.000000  ...  1599.000000    1599.000000  1599.000000   
mean      0.489847  ...     0.270976       2.538806     0.087467   
std       0.114126  ...     0.194801       1.409928     0.047065   
min       0.220000  ...     0.000000       0.900000     0.012000   
25%       0.410000  ...     0.090000       1.900000     0.070000   
50%       0.470000  ...     0.260000       2.200000     0.079000   
75%       0.550000  ...     0.420000       2.600000     0.090000   
max       1.080000  ...     1.000000      15.500000     0.611000   

                                                                          \
      free sulfur dioxide total sulfur dioxide      density           pH   
count         1599.000000          1599.000000  1599.000000  1599.000000   
mean            15.874922            46.467792     0.996747     3.311113   
std             10.460157            32.895324     0.001887     0.154386   
min              1.000000             6.000000     0.990070     2.740000   
25%              7.000000            22.000000     0.995600     3.210000   
50%             14.000000            38.000000     0.996750     3.310000   
75%             21.000000            62.000000     0.997835     3.400000   
max             72.000000           289.000000     1.003690     4.010000   

                                              
         sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  
mean      0.658149    10.422983     5.636023  
std       0.169507     1.065668     0.807569  
min       0.330000     8.400000     3.000000  
25%       0.550000     9.500000     5.000000  
50%       0.620000    10.200000     6.000000  
75%       0.730000    11.100000     6.000000  
max       2.000000    14.900000     8.000000  

[8 rows x 24 columns]

In [123]:
desc_white = white_copy.describe().transpose()
desc_red = red_copy.describe().transpose()
result = pd.concat([desc_white, desc_red], axis=1, keys=['white_copy', 'red_copy'])
display(result)

white_copy                                              \
                          count        mean        std      min         25%   
fixed acidity            4898.0    6.854788   0.843868  3.80000    6.300000   
volatile acidity         4898.0    0.278241   0.100795  0.08000    0.210000   
citric acid              4898.0    0.334192   0.121020  0.00000    0.270000   
residual sugar           4898.0    6.391415   5.072058  0.60000    1.700000   
chlorides                4898.0    0.045772   0.021848  0.00900    0.036000   
free sulfur dioxide      4898.0   35.308085  17.007137  2.00000   23.000000   
total sulfur dioxide     4898.0  138.360657  42.498065  9.00000  108.000000   
density                  4898.0    0.994027   0.002991  0.98711    0.991723   
pH                       4898.0    3.188267   0.151001  2.72000    3.090000   
sulphates                4898.0    0.489847   0.114126  0.22000    0.410000   
alcohol                  4898.0   10.514267   1.230621  8.00000    9.500000   
quality                  4898.0    5.877909   0.885639  3.00000    5.000000   

                                                     red_copy             \
                            50%       75%        max    count       mean   
fixed acidity           6.80000    7.3000   14.20000   1599.0   8.319637   
volatile acidity        0.26000    0.3200    1.10000   1599.0   0.527821   
citric acid             0.32000    0.3900    1.66000   1599.0   0.270976   
residual sugar          5.20000    9.9000   65.80000   1599.0   2.538806   
chlorides               0.04300    0.0500    0.34600   1599.0   0.087467   
free sulfur dioxide    34.00000   46.0000  289.00000   1599.0  15.874922   
total sulfur dioxide  134.00000  167.0000  440.00000   1599.0  46.467792   
density                 0.99374    0.9961    1.03898   1599.0   0.996747   
pH                      3.18000    3.2800    3.82000   1599.0   3.311113   
sulphates               0.47000    0.5500    1.08000   1599.0   0.658149   
alcohol                10.40000   11.4000   14.20000   1599.0  10.422983   
quality                 6.00000    6.0000    9.00000   1599.0   5.636023   

                                                                        \
                            std      min      25%       50%        75%   
fixed acidity          1.741096  4.60000   7.1000   7.90000   9.200000   
volatile acidity       0.179060  0.12000   0.3900   0.52000   0.640000   
citric acid            0.194801  0.00000   0.0900   0.26000   0.420000   
residual sugar         1.409928  0.90000   1.9000   2.20000   2.600000   
chlorides              0.047065  0.01200   0.0700   0.07900   0.090000   
free sulfur dioxide   10.460157  1.00000   7.0000  14.00000  21.000000   
total sulfur dioxide  32.895324  6.00000  22.0000  38.00000  62.000000   
density                0.001887  0.99007   0.9956   0.99675   0.997835   
pH                     0.154386  2.74000   3.2100   3.31000   3.400000   
sulphates              0.169507  0.33000   0.5500   0.62000   0.730000   
alcohol                1.065668  8.40000   9.5000  10.20000  11.100000   
quality                0.807569  3.00000   5.0000   6.00000   6.000000   

                                 
                            max  
fixed acidity          15.90000  
volatile acidity        1.58000  
citric acid             1.00000  
residual sugar         15.50000  
chlorides               0.61100  
free sulfur dioxide    72.00000  
total sulfur dioxide  289.00000  
density                 1.00369  
pH                      4.01000  
sulphates               2.00000  
alcohol                14.90000  
quality                 8.00000

In [134]:

for val in white_copy.columns:
    desc1 = white_copy[val].describe()
    desc2 = red_copy[val].describe()
    result = pd.concat([desc1, desc2], axis=1, keys=['white_copy', 'red_copy'])
    try:
        result['diff_%'] = ((result.iloc[:, 1] - result.iloc[:, 0])/result.iloc[:, 0])*100
    except:
        pass
    display(val)
    display(result.iloc[1:])

    

'fixed acidity'

white_copy   red_copy      diff_%
mean    6.854788   8.319637   21.369730
std     0.843868   1.741096  106.323246
min     3.800000   4.600000   21.052632
25%     6.300000   7.100000   12.698413
50%     6.800000   7.900000   16.176471
75%     7.300000   9.200000   26.027397
max    14.200000  15.900000   11.971831

'volatile acidity'

white_copy  red_copy      diff_%
mean    0.278241  0.527821   89.698961
std     0.100795  0.179060   77.648203
min     0.080000  0.120000   50.000000
25%     0.210000  0.390000   85.714286
50%     0.260000  0.520000  100.000000
75%     0.320000  0.640000  100.000000
max     1.100000  1.580000   43.636364

'citric acid'

white_copy  red_copy     diff_%
mean    0.334192  0.270976 -18.916069
std     0.121020  0.194801  60.966330
min     0.000000  0.000000        NaN
25%     0.270000  0.090000 -66.666667
50%     0.320000  0.260000 -18.750000
75%     0.390000  0.420000   7.692308
max     1.660000  1.000000 -39.759036

'residual sugar'

white_copy   red_copy     diff_%
mean    6.391415   2.538806 -60.277880
std     5.072058   1.409928 -72.202051
min     0.600000   0.900000  50.000000
25%     1.700000   1.900000  11.764706
50%     5.200000   2.200000 -57.692308
75%     9.900000   2.600000 -73.737374
max    65.800000  15.500000 -76.443769

'chlorides'

white_copy  red_copy      diff_%
mean    0.045772  0.087467   91.090320
std     0.021848  0.047065  115.421873
min     0.009000  0.012000   33.333333
25%     0.036000  0.070000   94.444444
50%     0.043000  0.079000   83.720930
75%     0.050000  0.090000   80.000000
max     0.346000  0.611000   76.589595

'free sulfur dioxide'

white_copy   red_copy     diff_%
mean   35.308085  15.874922 -55.038848
std    17.007137  10.460157 -38.495487
min     2.000000   1.000000 -50.000000
25%    23.000000   7.000000 -69.565217
50%    34.000000  14.000000 -58.823529
75%    46.000000  21.000000 -54.347826
max   289.000000  72.000000 -75.086505

'total sulfur dioxide'

white_copy    red_copy     diff_%
mean  138.360657   46.467792 -66.415459
std    42.498065   32.895324 -22.595712
min     9.000000    6.000000 -33.333333
25%   108.000000   22.000000 -79.629630
50%   134.000000   38.000000 -71.641791
75%   167.000000   62.000000 -62.874251
max   440.000000  289.000000 -34.318182

'density'

white_copy  red_copy     diff_%
mean    0.994027  0.996747   0.273564
std     0.002991  0.001887 -36.897603
min     0.987110  0.990070   0.299865
25%     0.991723  0.995600   0.390986
50%     0.993740  0.996750   0.302896
75%     0.996100  0.997835   0.174179
max     1.038980  1.003690  -3.396601

'pH'

white_copy  red_copy    diff_%
mean    3.188267  3.311113  3.853083
std     0.151001  0.154386  2.242286
min     2.720000  2.740000  0.735294
25%     3.090000  3.210000  3.883495
50%     3.180000  3.310000  4.088050
75%     3.280000  3.400000  3.658537
max     3.820000  4.010000  4.973822

'sulphates'

white_copy  red_copy     diff_%
mean    0.489847  0.658149  34.358077
std     0.114126  0.169507  48.526389
min     0.220000  0.330000  50.000000
25%     0.410000  0.550000  34.146341
50%     0.470000  0.620000  31.914894
75%     0.550000  0.730000  32.727273
max     1.080000  2.000000  85.185185

'alcohol'

white_copy   red_copy     diff_%
mean   10.514267  10.422983  -0.868191
std     1.230621   1.065668 -13.404049
min     8.000000   8.400000   5.000000
25%     9.500000   9.500000   0.000000
50%    10.400000  10.200000  -1.923077
75%    11.400000  11.100000  -2.631579
max    14.200000  14.900000   4.929577

'quality'

white_copy  red_copy     diff_%
mean    5.877909  5.636023  -4.115185
std     0.885639  0.807569  -8.815011
min     3.000000  3.000000   0.000000
25%     5.000000  5.000000   0.000000
50%     6.000000  6.000000   0.000000
75%     6.000000  6.000000   0.000000
max     9.000000  8.000000 -11.111111

'quality_label'

white_copy red_copy
unique          3        3
top        medium   medium
freq         4535      837

<span style="font-size: larger; color: green;">Wine Comparisons</span><br>
<span style="color: green;">Fixed acidity:</span> 21% difference in mean. red > white. Red also has a larger spread.<br>
<span style="color: green;">Volatile acidity:</span>  red > white 89% bigger. significantly bigger spread<br>
<span style="color: green;">Citric Acid:</span> White is 20% more acidic than red<br>
<span style="color: green;">Residual Sugar:</span> White has quite a lot more sugar 60% difference<br>
<span style="color: green;">Chlorides:</span> red has almost twice as many chlorides as white<br>
<span style="color: green;">Free sulfur dioxide:</span> large difference. White has more than twice as many as red does.<br>
<span style="color: green;">Total sulfur dioxide:</span> even bigger diffence. white is greater by a multiple of 3><br>
<span style="color: green;">Density:</span> no difference<br>
<span style="color: green;">PH:</span> no real difference. 4%<br>
<span style="color: green;">Sulphates:</span> red has 34% more sulphates. larger spread too<br>
<span style="color: green;">alcohol:</span> no difference<br>
<span style="color: green;">qualtiy:</span> no difference<br>

<span style= "font-size: 32px; color: green; font-family: 'arial';">TIME FOR PLOTS</span>

lets make a pie chart comparing numbers of red and white

In [148]:
# pie_labels = {'Red': red_copy.shape[0], 'White': white_copy.shape[0]}
# print(pie_labels)
labels = ['White', 'Red']
sizes = [len(white_copy), len(red_copy)]

print("Matplotlib version:", matplotlib.__version__)

plt.pie(sizes, labels=labels, autopct='%1.1f%%', startangle=90)
plt.title('Red or White?')



NameError: name 'matplotlib' is not defined